monkeysphere 0.44_2

bondolo opened the pull request:
- [x] Have you followed the [guidelines for contributing](https://github.com/Homebrew/homebrew-core/blob/HEAD/CONTRIBUTING.md)?
- [x] Have you checked that there aren't other open [pull requests](https://github.com/Homebrew/homebrew-core/pulls) for the same formula update/change?
- [x] Have you built your formula locally with `brew install --build-from-source <formula>`, where `<formula>` is the name of the formula you're submitting?
- [x] Is your test running fine `brew test <formula>`, where `<formula>` is the name of the formula you're submitting?
- [x] Does your build pass `brew audit --strict <formula>` (after doing `brew install <formula>`)?

-----

The change adds two additional dependencies discovered through use on a new "fresh" system. Commands that had worked elsewhere failed on the new system because I had not installed undeclared dependencies. Once bash and coreutils were installed everything worked better.

bondolo commented:
I feel somewhat mixed about the removing bash dependency request. On one hand I understand and agree except that if the user has requested bash as login shell they will still get Apple's bash in a terminal and on ssh as the full path of the shell in /etc/passwd or /etc/shells

For me compatibility with linux scripts and utils is important so I always install bash to get a current version for much better compatibility. As pointed out, zsh is now the default for users who didn't know otherwise anyway. Adding the homebrew bash package would change the shell used if a user types 'bash' in to a shell or runs a script with a ```#! bash``` bang path. I suspect there are few scripts written specifically for Apple's bash--they are either going to be attempts at posix sh scripts or generic bash.

bondolo commented:
Perhaps add a caveat? The only other alternatives are to install a keg-only bash or to delete monkeysphere as it can't run with the default Apple provided bash.

MikeMcQuaid commented on Formula/monkeysphere.rb:
```diff
@@ -15,6 +16,8 @@ class Monkeysphere < Formula
   end
 
   depends_on "gnu-sed" => :build
+  depends_on "bash" # Apple's BASH 3.2 is insufficient, BASH 4.X features are used
```
This doesn't swap out their default shell unless they configure it to do so.

bayandin replied:
> This doesn't swap out their default shell unless they configure it to do so.

Actually, this is true, it will not change the user shell (I've tried it on both Catalina and Mojave), sorry for the confusion.

Nevertheless, it will change a result for `/usr/bin/env bash`.

reitermarkus replied:
I think `bash` 4 is backwards compatible, so the only case where this would be an issue is if you are developing a script which should run on system `bash` and you accidentally introduce newer features due to this.

MikeMcQuaid replied:
This is an argument for a `keg_only` `bash`.

reitermarkus replied:
I thought `bash` was *not* `keg_only` for a reason. Or do you mean adding a versioned `keg_only` `bash@4`?

bondolo replied:
If the desire is to allow packages like Monkeysphere which require bash 4.0+ without changing the bash which will be invoked on the from the path then some form of keg only bash will be needed. Making the existing bash formula keg_only would make a lot of people, like me, unhappy because we are already using it as our primary shell for better Linux compatibility.

MikeMcQuaid replied:
> I thought `bash` was _not_ `keg_only` for a reason. Or do you mean adding a versioned `keg_only` `bash@4`?

@reitermarkus These decisions are all somewhat arbitrary. I'd happily see all system duplicates be keg-only, personally.

We're not going to add multiple versions of Bash with different statuses.



> Making the existing bash formula keg_only would make a lot of people, like me, unhappy because we are already using it as our primary shell for better Linux compatibility.

You can add a single path to your PATH and it'd work the same as before.

bondolo replied:
With the most recent changes the monkeysphere formula no longer cares whether bash is on the path or is keg only. This seems sufficient for this PR.

Can the decisions about whether to make bash formula keg_only and/or add a caveat to the bash formula about path alteration be handled on a separate issue/PR? I have wanted to get more involved with homebrew (see various cask version updates, adding Webots cask, this PR, etc. ) but decisions about bash seem to be out of scope even if this PR did initiate the discussion.

bondolo commented:
I've added a caveat and changed the shebang for Monkeysphere scripts in case bash is installed key-only.

If bash can't be specified as a dependency then Monkeysphere formula should be removed because it is broken and incompatible with Apple bash.

SMillerDev commented:
You changed 2300 files, please rebase on top of the main branch with only your changes.

bondolo commented:
Now correctly rebased on upstream master

reitermarkus commented on Formula/monkeysphere.rb:
```diff
@@ -1,16 +1,22 @@
     regex(/href=.*?monkeysphere.?v?(\d+(?:\.\d+)+)(?:\.orig)?\.t/i)
   end
 
+  livecheck do
```
This is now duplicated.

bondolo replied:
updated to resolve this

SMillerDev commented on Formula/monkeysphere.rb:
```diff
@@ -48,6 +53,35 @@ def install
+
+    # Needs BASH 4.0 or later, Apple's BASH 3.2 is insufficient
+    inreplace bin/"monkeysphere", "#!/usr/bin/env bash",
+                                  "#!/usr/local/opt/bash/bin/bash"
```
This should be `#{Formula["bash"].opt_bin}/bash`

bondolo replied:
I will make this change.

SMillerDev commented on Formula/monkeysphere.rb:
```diff
@@ -48,6 +53,35 @@ def install
+
+  def caveats
+    <<~EOS
+      This formula installs BASH as a dependency because Monkeysphere requires a newer 
```
I think this should be a caveat in the bash formula or not at all.

bondolo replied:
I think it is worth repeating in this formula and anything else that has bash as a dependency as people may not notice the caveat streaming by on the dependency install.

SMillerDev replied:
Caveats are gathered at the end of all install output, I don't think they're gonna read this one if they won't read the bash one.

SMillerDev commented on Formula/monkeysphere.rb:
```diff
@@ -48,6 +53,35 @@ def install
+    inreplace pkgshare/"common", "$(find ",
+                                 "$(gfind "
+    # use liblockfile
+    inreplace pkgshare/"common", "lockfile-create",
```
These should have patches upstreamed.

bondolo replied:
I have already reported these issues upstream. I reported them in their irc channel since their bug tracker seemed to be dead.

SMillerDev commented on Formula/monkeysphere.rb:
```diff
@@ -20,10 +21,14 @@ class Monkeysphere < Formula
   depends_on "gnu-sed" => :build
+  depends_on "bash" # Apple's BASH 3.2 is insufficient, BASH 4.X features are used
+  depends_on "coreutils"
+  depends_on "findutils" # GNU extensions are used with find
```
Only during the build right?

bondolo replied:
No, these are installed along with the monkeysphere formula on user's machines.

SMillerDev reviewed (commented):
Any response on irc about the homepage?

bondolo commented:
> Any response on irc about the homepage?

None, though the channel is archived and there were a couple of the devs idling so presumably there is some chance somebody saw it. The bug tracker appeared to be unused since 2017.

MikeMcQuaid reviewed (changes requested):
I'm increasingly wondering if this formula is maintained enough to be worth this upgrade, unfortunately. Bad signs:

- https homepage doesn't work
- won't work with macOS system shell and needs patched
- needs a bunch of GNU tools (`gnu-sed`, `coreutils`, `findutils`) that start to look like it's not really been ported to macOS/BSD at all
- requires a bunch of `inreplace`s

bondolo commented:
> I'm increasingly wondering if this formula is maintained enough to be worth this upgrade, unfortunately. Bad signs:
>     * https homepage doesn't work
> 
>     * won't work with macOS system shell and needs patched
> 
>     * needs a bunch of GNU tools (`gnu-sed`, `coreutils`, `findutils`) that start to look like it's not really been ported to macOS/BSD at all
> 
>     * requires a bunch of `inreplace`s
> 
>     * archived IRC channel
> 
>     * bug tracker appeared unused since 2017.

One note: The "archived IRC" referred to the channel being recorded with the channel logs being available online, not that the channel is in any way defunct. 

Agreed that it is not the most desirable form for a package. With the patches it works correctly for me and I will continue to use it. Presumably it might work just as well for others. (I actually make one additional change to the scripts for my use; adding the Apple specific "-K" option to ssh-add invocation to cause private keys to be added to MacOS Keychain).

Only by this issue lingering so long did I take the time to improve the "MacOS Port" of the package. I was/am unsure how much porting work is normally allowed in Homebrew formula.

MikeMcQuaid commented:
> One note: The "archived IRC" referred to the channel being recorded with the channel logs being available online, not that the channel is in any way defunct.

Thanks for clarifying 👍🏻 



> Only by this issue lingering so long did I take the time to improve the "MacOS Port" of the package. I was/am unsure how much porting work is normally allowed in Homebrew formula.

Yeh, normally: not much unfortunately 😭. This could be a good candidate for you to run a tap, though.

@homebrew/core any thoughts on accepting this PR/deprecating/disabling this formula?

vszakats commented:
Instead of reverting to the insecure homepage, can we notify the webpage maintainer about their expired Let's Encrypt certificate, so they can fix it?

SMillerDev commented:
> Instead of reverting to the insecure homepage, can we notify the webpage maintainer about their expired Let's Encrypt certificate, so they can fix it?

That was done here: https://github.com/Homebrew/homebrew-core/pull/59949#issuecomment-678494510. They're not responding

SMillerDev commented:
> any thoughts on accepting this PR/deprecating/disabling this formula?

I'd say it should be disabled without the bash update that way it can be taken over easily in a tap

bondolo commented:
Any final thoughts? Approve this change or remove Monkeysphere formula?

MikeMcQuaid commented:
> I'd say it should be disabled without the bash update that way it can be taken over easily in a tap

Agreed and given no-one disagrees I think that's the way forward. Sorry @bondolo!

reitermarkus commented:
There are a bunch of other formulae depending on `bash` already, so how is this different? Do we disable all others as well?

MikeMcQuaid commented:
@reitermarkus There's many more issues here than just depending on Bash, see above.

bondolo commented:
> > I'd say it should be disabled without the bash update that way it can be taken over easily in a tap
> 
> Agreed and given no-one disagrees I think that's the way forward. Sorry @bondolo!

Not a problem. I am glad that this issue was able to bring to light some of the bash dependency issues. 

Closing this PR. Someone else on the homebrew team can do the PR to deprecate/remove Monkeysphere formula.

I may, as you suggest, create a tap for this issue. The Linux Homebrew project may also be interested in keeping this formula since they would require none of the gross hacks.
